src/main/java/org/candlepin/subscriptions/product/MidProduct.java

```diff
@@ -0,0 +1,376 @@
+    SOCKET_LIMIT,
+    SERVICE_TYPE,
+    PRODUCT_FAMILY,
+    PRODUCT_NAME,
+    USAGE,
+    /** Originates from opProd attr code "CORES" when coming from a derived product. */
+    __X_VIRTUAL_CORES,
+    /** Originates from opProd attr code "IFL" when coming from a derived product. */
+    __X_VIRTUAL_IFL,
+    /** Originates from opProd attr code "SOCKET_LIMIT" when coming from a derived product. */
+    __X_VIRTUAL_SOCKETS,
+    /** Role originates from opProd field role, not an attribute. */
+    __X_ROLE;
+  }
+
+  /** Maps opProd attribute codes to MidProduct.Attrs of the same name. */
+  private static final Map<String, Attr> CODE_TO_ENUM =
+      Arrays.stream(Attr.values())
+          .collect(Collectors.toMap(Attr::name, attr -> attr, (a1, a2) -> a1));
+
+  private MidProduct(String sku) {
+    this.sku = sku;
+  }
+
+  private String sku;
+  private SortedSet<String> children = new TreeSet<>();
+  private SortedSet<Integer> engOids = new TreeSet<>();
+  private EnumMap<Attr, String> attrs = new EnumMap<>(Attr.class);
+  private StringBuilder conflicts = new StringBuilder(0);
+
+  /**
+   * Create an {@link Offering} based on product service data from upstream.
+   *
+   * @param sku the identifier of the marketing operational product
+   * @param productService the upstream product service
+   * @return An Offering with information filled by an upstream service, or empty if the product was
+   *     not found.
+   */
+  public static Optional<Offering> offeringFromUpstream(String sku, ProductService productService) {
+    LOGGER.debug("Retrieving product tree for offering sku=\"{}\"", sku);
+
+    try {
+      return productService
+          .getTree(sku)
+          .map(MidProduct::createFromTree)
+          .map(mid -> mid.fetchAndAddDerivedTreeIfExist(productService))
+          .map(mid -> mid.fetchAndAddEngProdsIfExist(productService))
+          .map(MidProduct::toOffering);
+    } catch (ApiException e) {
+      throw new RuntimeException("Unable to retrieve upstream offering sku=\"" + sku + "\"", e);
+    }
+  }
+
+  private static MidProduct createFromTree(RESTProductTree skuTree) {
+    var products = skuTree.getProducts();
+
+    // Product trees returned by the product service always list the parent product first.
+    var parent = products.get(0);
+    var children = products.subList(1, products.size());
+
+    var offer = new MidProduct(parent.getSku());
+    // Though theoretically possible, none of the products in use by Candlepin (which includes the
+    // allowlisted swatch products) ever has more than one role.
+    String role = parent.getRoles().stream().findFirst().orElse(null);
+    offer.attrs.put(Attr.__X_ROLE, role);
+    offer.mapAttributes(parent);
+
+    // For each child, merge its unconflcting information into the parent.
+    children.stream().map(MidProduct::createFromProduct).forEach(offer::merge);
+
+    return offer;
+  }
+
+  private Offering toOffering() {
+    if (conflicts.length() != 0) {
```

Comment:
When the initial convert-into-Offering logic was written, I wanted to make sure that the "merge attrs+fields from multiple products into a single Offering" logic made sense. I wrote it so it would throw an exception whenever two or more products had different values for the same attr or field.

As it is now written, it will still collect those collisions, but it will only log the collisions, and will create an Offering regardless.

If this seems like it won't be useful, let me know and I can remove it. It is straightforward enough to remove.